### DISCUSSION TITLE
How am I supposed to retry P1017: Server has closed the connection

### DISCUSSION THREAD

**Discussion by jeff-an** [prisma/prisma#24490] [Q&A]
### Question

Possibly related to [this issue](https://github.com/prisma/prisma/issues/9902), but I'm confused as to how applications are supposed to gracefully handle transient DB connection errors due to things like database/pgbouncer restarts. It seems like queries like `prisma.$executeRaw`SELECT 1;`` will just fail immediately with P1017 if the db connection is severed during the query. I would expect that in these cases prisma automatically attempts to reconnect and retry the query, but it doesn't appear to do that. I'm using pgbouncer in transaction pooling mode if that makes a difference.

What is the recommended way to handle these errors? Is it really just retrying within application code? If so, that seems incredibly suboptimal to have to wrap every `prisma` call, and this is not something suggested or recommended in the docs either.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @jeff-an 👋 

Thank you for raising this question.

To better assist with this issue,  Can you please fill out the template with specific details, including the Prisma version and the output of `npx prisma -v`. Without this information, it is difficult to diagnose the problem accurately. Additionally, it would be helpful to know your connection string options. For example, whether `sslmode=verify-full` or other specific settings are applied in the connection string.

**Reply by ludralph** [reply]
In general, the user needs to handle transient database connection errors on their end, as Prisma does not retry individual queries on connection failure. This error should be rare but can occur during PgBouncer restarts. Using "reload" instead of restarting PgBouncer or having multiple instances can mitigate this issue. see `How to upgrade PgBouncer without dropping connections?`
https://www.pgbouncer.org/faq.html.

Additionally, we would be interested in knowing if other libraries, database clients, or ORMs handle this differently, as it could provide valuable insights.

**Reply by jeff-an** [reply]
Hey there!

Graceful upgrades don't really help as entire pods (we use kubernetes) can churn for arbitrary reasons, such as the underlying node being replaced, the pod failing a healthcheck, or transient network issues. In these cases, the clients are going to have momentary disconnections and no amount of graceful handling on the pgbouncer side would help.

Sequelize apparently supports a "retries" option that lets you automatically retry queries if they fail because of a certain error: https://github.com/sequelize/website/issues/167 However it's not well documented either.

**Reply by jeff-an** [reply]
I'm going to retry the following paradigm and see if it works:

```
client.$use(async (params, next) => {
    let attempts = 1;

    while (true) {
      attempts++;
      try {
        const result = await next(params);
        return result;
      } catch (err) {
        if (attempts >= MAX_ATTEMPTS) {
          throw err;
        }
        if (
          (err instanceof PrismaClientKnownRequestError &&
            retryableErrorCodes.includes(err.code)) ||
          (err instanceof PrismaClientInitializationError &&
            retryableErrorCodes.includes(err.errorCode ?? ""))
        ) {
          console.warn(`Retrying query due to error: ${err.message}`);
          continue;
        }
        throw err;
      }
    }
  });
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
